plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpooledSegmentHandle.java

```diff
@@ -13,42 +13,94 @@
 
-public record FileSystemSpooledSegmentHandle(String objectName, Instant validUntil, Optional<Slice> encryptionKey)
+public record FileSystemSpooledSegmentHandle(QueryId queryId, byte[] ulid, Optional<Slice> encryptionKey)
         implements SpooledSegmentHandle
 {
+    private static final String OBJECT_NAME_SEPARATOR = "::";
+
     public FileSystemSpooledSegmentHandle
     {
-        requireNonNull(objectName, "objectName is null");
-        validUntil = requireNonNull(validUntil, "validUntil is null").truncatedTo(ChronoUnit.MILLIS);
+        requireNonNull(queryId, "queryId is null");
+        requireNonNull(encryptionKey, "encryptionKey is null");
+        verify(ulid.length == 16, "ULID must be 128 bits");
     }
 
-    public static FileSystemSpooledSegmentHandle random(SpoolingContext context, Duration ttl)
+    public static FileSystemSpooledSegmentHandle random(SecureRandom random, QueryId queryId, Instant expireAt)
     {
-        return random(context, ttl, Optional.empty());
+        return random(random, queryId, expireAt, Optional.empty());
     }
 
-    public static FileSystemSpooledSegmentHandle random(SpoolingContext context, Duration ttl, Optional<Slice> encryptionKey)
+    public static FileSystemSpooledSegmentHandle random(SecureRandom random, QueryId queryId, Instant expireAt, Optional<Slice> encryptionKey)
     {
         return new FileSystemSpooledSegmentHandle(
-                context.queryId().getId() + "/" + randomObjectName(),
-                Instant.now().plusMillis(ttl.toMillis()),
+                queryId,
+                ULID.generateBinary(expireAt.truncatedTo(MILLIS).toEpochMilli(), entropy(random)),
                 encryptionKey);
     }
 
-    private static String randomObjectName()
+    public Instant expirationTime()
+    {
+        return Instant.ofEpochMilli(ULID.getTimestampBinary(ulid()));
+    }
+
+    // Storage object name starts with the ULID which is ordered lexicographically
+    // by the time of the expiration, which makes it possible to find the expired
+    // segments by listing the storage objects. This is crucial for the storage
+    // cleanup process to be able to delete the expired segments.
+    //
+    // See for the ULID spec: https://github.com/ulid/spec
+    public String storageObjectName()
+    {
+        return ULID.fromBinary(ulid) + '@' + queryId;
+    }
+
+    public static FileSystemSpooledSegmentHandle fromStorageObjectName(String objectName)
+    {
+        List<String> parts = Splitter.on(OBJECT_NAME_SEPARATOR).limit(2).splitToList(objectName);
+        byte[] ulid = ULID.toBinary(parts.get(0));
+        QueryId queryId = QueryId.valueOf(parts.get(1));
+
+        return new FileSystemSpooledSegmentHandle(queryId, ulid, Optional.empty());
+    }
+
+    public static FileSystemSpooledSegmentHandle of(QueryId queryId, byte[] ulid, Optional<Slice> encryptionKey)
+    {
+        verify(ULID.isValidBinary(ulid), "Invalid ulid: %s", ULID.fromBinary(ulid));
```

Comment:
This will convert to string every time. Also, we shouldn't call `ULID.fromBinary()` when it is not valid binary. So this should be
```java
if (!ULID.isValidBinary(ulid)) {
    throw new IllegalArgumentException("Invalid ULID: " + HexFormat.of().formatHex(ulid));
}